src/web100srv.c

```diff
@@ -1527,647 +1326,364 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
-      }
-
-      tmp_ptr = mchild;
-
-      // Update queued clients, send message to client when it moves
-      //  up in the queue enough to get closer to running a test.  This
-      //  happens when the client falls into the next lower maxqueue bin
-
-      if (waiting > (2 * max_clients)) {
-        int wait_data = -1;
-        int num_secs = 0;
-        for (i = max_clients; i <= waiting; i++) {
-          if (tmp_ptr == NULL) break;
-          if (i == (2 * max_clients)) {
-            wait_data = 1;
-            num_secs = 45;
-          }
-          if (i == (3 * max_clients)) {
-            wait_data = 2;
-            num_secs = 90;
-          }
-          if (wait_data != -1) {
-            child_signal = wait_data;
-            write(tmp_ptr->pipe, &child_signal, sizeof(int));
-            log_println(6,
-                        "sent %d sec update message to client %d on fd=%d, "
-                        "send_msg() returned %d",
-                        num_secs, tmp_ptr->pid, tmp_ptr->pipe);
-          }
-          tmp_ptr = tmp_ptr->next;
-        }
-      }
-      goto dispatch_client;
-    }
-
-    if ((waiting < 0) || (mclients < 0)) {
-      log_println(6,
-                  "Fault: Negative number of clents waiting=%d, mclients=%d"
-                  ", nuke them",
-                  waiting, mclients);
-      while (head_ptr != NULL) {
-        // give message to server waiting in queue
-        child_signal = SRV_QUEUE_SERVER_BUSY;
-        write(head_ptr->pipe, &child_signal, sizeof(int));
-        close(head_ptr->pipe);
-        tpid = head_ptr->pid;
-        child_sig(-1);
-        kill(tpid, SIGTERM);
-        child_sig(tpid);
-      }
-      waiting = 0;
-      mclients = 0;
-      continue;
-    }
-
-    if ((waiting == 0) && (head_ptr != NULL)) {
-      log_println(6, "Fault: Something [%d] in queue, but no waiting clients",
-                  head_ptr->pid);
-      while (head_ptr != NULL) {
-        /* send_msg(head_ptr->ctlsockfd, SRV_QUEUE, "9977", 4); */
-        child_signal = SRV_QUEUE_SERVER_BUSY;
-        write(head_ptr->pipe, &child_signal, sizeof(int));
-        close(head_ptr->pipe);
-        tpid = head_ptr->pid;
-        child_sig(-1);
-        kill(tpid, SIGTERM);
-        child_sig(tpid);
-      }
-      waiting = 0;
-      mclients = 0;
-      continue;
+  FD_ZERO(&rfd);
+  FD_SET(listenfd, &rfd);
+  FD_SET(signalfd, &rfd);
+  if (waitforever) {
+    retcode = select(max(listenfd, signalfd) + 1, &rfd, NULL, NULL, NULL);
+  } else {
+    sel_tv.tv_sec = 3;  // 3 seconds == WAIT_TIME_SRVR
+    sel_tv.tv_usec = 0;
+    retcode = select(max(listenfd, signalfd) + 1, &rfd, NULL, NULL, &sel_tv);
+  }
+  if (retcode == -1) {
+    if (errno != EINTR) {
+      log_println(0, "Error in server's select call: %d (%s)", errno,
+                  strerror(errno));
     }
+    return NULL;
+  }
+  // If we woke up due to data on the signalfd, then read the data.
+  if (FD_ISSET(signalfd, &rfd)) {
```

Comment:
Receiving data on the signalfd channel means "wake up and do work", not "a child process called exit()".  Right now, it is true that the only event that wakes up the server is SIGCHLD, but if we don't need to care about whether we received SIGCHLD or not, then I would prefer that we not care. It doesn't matter why the server woke up, it matters that the server woke up.  Upon wakeup: open a new connection, if requested; add the new client to the queue, if needed; and perform queue maintenance tasks.  None of these tasks change if the server woke up due to SIGCHLD.